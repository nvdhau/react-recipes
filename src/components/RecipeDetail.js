import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import classNames from 'classnames';

// classNames: deal with case props.className = null
const RecipeDetail = ({ recipe, style, className }) => {
  // zero state element
  if (!recipe) {
    return (
      <p
        style={style}
        className={classNames('h3 p2 bg-white italic center', className)}
      >
        Please select a recipe
      </p>
    );
  }

  // have data element
  return (
    <div style={style} className={classNames('p2 bg-white', className)}>
      <h2 className="h2">{recipe.name}</h2>
      <img className="fit" src={recipe.image} alt={recipe.name} />
      <div>
        <span>{recipe.category}</span>
        <span>{recipe.calories}</span>
      </div>
      <h3>Ingredients</h3>
      <ul>
        {recipe.ingredients.map(ingredient => (
          <li key={ingredient}>{ingredient}</li>
        ))}
      </ul>
      <h3>Steps</h3>
      <ol>
        {recipe.steps.map(step => (
          <li key={step}>{step}</li>
        ))}
      </ol>
      <Link to={`/recipe/${recipe.id}`}>See more</Link>
    </div>
  );
};

RecipeDetail.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  recipe: PropTypes.object,
};

export default RecipeDetail;
