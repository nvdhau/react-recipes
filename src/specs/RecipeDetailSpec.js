import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import RecipeDetail from '../components/RecipeDetail';

// think component as IO function
describe('<RecipeDetail />', () => {
  let testRecipe;

  beforeEach(() => {
    testRecipe = {
      id: 1,
      name: 'Test Recipe',
      category: 'Test Category',
      ingredients: ['Ing 1', 'Ing 2', 'Ing 3'],
      steps: ['Step 1', 'Step 2', 'Step 3'],
    };
  });

  test('Should render zero state', () => {
    // because RecipeDetail has Link, and Link needs to be inside BrowserRouter
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should correctly render', () => {
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should correctly assign a className', () => {
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} className="classname-test" />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should render recipe without ingredients', () => {
    delete testRecipe.ingredients;
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should render recipe without steps', () => {
    delete testRecipe.steps;
    const component = renderer.create(
      <BrowserRouter>
        <RecipeDetail recipe={testRecipe} />
      </BrowserRouter>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
