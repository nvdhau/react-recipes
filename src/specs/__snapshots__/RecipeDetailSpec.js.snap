// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeDetail /> Should correctly assign a className 1`] = `
<div
  className="p2 bg-white classname-test"
>
  <h2
    className="h2"
  >
    Test Recipe
  </h2>
  <img
    alt="Test Recipe"
    className="fit"
  />
  <div>
    <span>
      Test Category
    </span>
    <span />
  </div>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      Ing 1
    </li>
    <li>
      Ing 2
    </li>
    <li>
      Ing 3
    </li>
  </ul>
  <h3>
    Steps
  </h3>
  <ol>
    <li>
      Step 1
    </li>
    <li>
      Step 2
    </li>
    <li>
      Step 3
    </li>
  </ol>
  <a
    href="/recipe/1"
    onClick={[Function]}
  >
    See more
  </a>
</div>
`;

exports[`<RecipeDetail /> Should correctly render 1`] = `
<div
  className="p2 bg-white"
>
  <h2
    className="h2"
  >
    Test Recipe
  </h2>
  <img
    alt="Test Recipe"
    className="fit"
  />
  <div>
    <span>
      Test Category
    </span>
    <span />
  </div>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      Ing 1
    </li>
    <li>
      Ing 2
    </li>
    <li>
      Ing 3
    </li>
  </ul>
  <h3>
    Steps
  </h3>
  <ol>
    <li>
      Step 1
    </li>
    <li>
      Step 2
    </li>
    <li>
      Step 3
    </li>
  </ol>
  <a
    href="/recipe/1"
    onClick={[Function]}
  >
    See more
  </a>
</div>
`;

exports[`<RecipeDetail /> Should render recipe without ingredients 1`] = `
<div
  className="p2 bg-white"
>
  <h2
    className="h2"
  >
    Test Recipe
  </h2>
  <img
    alt="Test Recipe"
    className="fit"
  />
  <div>
    <span>
      Test Category
    </span>
    <span />
  </div>
  <h3>
    Ingredients
  </h3>
  <ul />
  <h3>
    Steps
  </h3>
  <ol>
    <li>
      Step 1
    </li>
    <li>
      Step 2
    </li>
    <li>
      Step 3
    </li>
  </ol>
  <a
    href="/recipe/1"
    onClick={[Function]}
  >
    See more
  </a>
</div>
`;

exports[`<RecipeDetail /> Should render recipe without steps 1`] = `
<div
  className="p2 bg-white"
>
  <h2
    className="h2"
  >
    Test Recipe
  </h2>
  <img
    alt="Test Recipe"
    className="fit"
  />
  <div>
    <span>
      Test Category
    </span>
    <span />
  </div>
  <h3>
    Ingredients
  </h3>
  <ul>
    <li>
      Ing 1
    </li>
    <li>
      Ing 2
    </li>
    <li>
      Ing 3
    </li>
  </ul>
  <h3>
    Steps
  </h3>
  <a
    href="/recipe/1"
    onClick={[Function]}
  >
    See more
  </a>
</div>
`;

exports[`<RecipeDetail /> Should render zero state 1`] = `
<p
  className="h3 p2 bg-white italic center"
>
  Please select a recipe
</p>
`;
